include/sparrow/any_data_utils.hpp

```diff
@@ -0,0 +1,159 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <ranges>
+#include <vector>
+
+#include "sparrow/memory.hpp"
+#include "sparrow/mp_utils.hpp"
+
+
+namespace sparrow
+{
+    /**
+     * Get a raw pointer from a smart pointer, a range, an object or a pointer.
+     *
+     * @tparam T The type of the pointer.
+     * @tparam U The type of the element.
+     * @param elem The element.
+     * @return A raw pointer.
+     */
+    template <typename T, typename U>
+    T* get_raw_ptr(U& elem)
+    {
+        if constexpr (std::is_pointer_v<U>)
+        {
+            return reinterpret_cast<T*>(elem);
+        }
+        else if constexpr (mpl::has_element_type<U>)
+        {
+            if constexpr (mpl::smart_ptr<U> || std::derived_from<U, std::shared_ptr<typename U::element_type>>
+                          || mpl::is_type_instance_of_v<U, value_ptr>)
+            {
+                if constexpr (std::same_as<typename U::element_type, T> || std::same_as<T, void>)
+                {
+                    return reinterpret_cast<T*>(elem.get());
+                }
+                else if constexpr (mpl::has_data_function<typename U::element_type, T>)
+                {
+                    return reinterpret_cast<T*>(std::ranges::data(*elem.get()));
+                }
+            }
+        }
+        else if constexpr (std::ranges::input_range<U>)
+        {
+            return reinterpret_cast<T*>(std::ranges::data(elem));
+        }
+        else if constexpr (mpl::has_data_function<U, T>)
+        {
+            return reinterpret_cast<T*>(elem.data());
+        }
+        else if constexpr (std::same_as<T, U>)
+        {
+            return reinterpret_cast<T*>(&elem);
+        }
+        else if constexpr (std::same_as<T, void>)
+        {
+            return reinterpret_cast<void*>(&elem);
+        }
+        else
+        {
+            static_assert(mpl::dependent_false<T, U>::value, "get_raw_ptr: unsupported type.");
+            mpl::unreachable();
+        }
+    }
+
+    /**
+     * Create a vector of pointers from a range.
+     *
+     * @tparam T The type of the pointers.
+     * @tparam Range The range type.
+     * @tparam Allocator The allocator type.
+     * @param range The range.
+     * @return A vector of pointers.
+     */
+    template <class T, std::ranges::input_range Range, template <typename> class Allocator = std::allocator>
+    std::vector<T*, Allocator<T*>> to_raw_ptr_vec(Range& range)
+    {
+        std::vector<T*, Allocator<T*>> raw_ptr_vec;
+        raw_ptr_vec.reserve(range.size());
+        std::ranges::transform(
+            range,
+            std::back_inserter(raw_ptr_vec),
+            [](auto& elem) -> T*
+            {
+                return get_raw_ptr<T>(elem);
+            }
+        );
+        return raw_ptr_vec;
+    }
+
+    /**
+     * Create a vector of pointers from a tuple.
```

Comment:
```suggestion
     * Create a vector of pointers to elements of a tuple.
```

Reply:
Does that mean the tuple is required to always have elements of the same type? Is that still a tuple?

Reply:
Yes the types can be different. I added documentation about that.